File: JTalk/src/Models/ComplexMessage.java
package Models;

import java.util.Date;
import java.util.List;
import java.util.Map;

public class ComplexMessage {

    public Date Timestamp;

    // Gson gives back a LinkedList for C# System.Collections.Generic.List
    public List<Message> Messages;

    // Gson gives back a LinkedHashMap for a .NET Dictionary<string, string>
    public Map<String, String> Metadata;




    public String toString() {
        StringBuilder builder = new StringBuilder();

        if (Messages != null) {
            builder.append("Messages : \n");
            for (Message nestedMessage : Messages)
                builder.append("\t" + nestedMessage.Number + " : " + nestedMessage.Text);
        }
        builder.append("\nTimeStamp : " + Timestamp + "\n");
        if (Metadata != null) {
            builder.append("Metadata : \n");
            for (Map.Entry<String, String> pair : Metadata.entrySet())
                builder.append("\t" + pair.getKey() + " : " + pair.getValue() + "\n");
        }
        return builder.toString();
    }
}

File: TalkSharp/TalkSharp/Models/ComplexMessage.cs
using System;
using System.Collections.Generic;

namespace TalkSharp.Models
{
    public class ComplexMessage
    {
        public ComplexMessage(DateTime Timestamp, Dictionary<string, string> Metadata, List<Message> Messages)
        {
            this.Timestamp = Timestamp;
            this.Metadata = Metadata;
            this.Messages = Messages;
        }

        public DateTime Timestamp { get; set; }

        public Dictionary<string, string> Metadata { get; set; }

        public List<Message> Messages { get; set; }
    }
}